// react-native-paper/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Colors: true,
  useTheme: true,
  withTheme: true,
  ThemeProvider: true,
  Provider: true,
  DefaultTheme: true,
  DarkTheme: true,
  shadow: true,
  overlay: true,
  configureFonts: true,
  Avatar: true,
  List: true,
  Drawer: true,
  AnimatedFAB: true,
  Badge: true,
  ActivityIndicator: true,
  Banner: true,
  BottomNavigation: true,
  Button: true,
  Card: true,
  Checkbox: true,
  Chip: true,
  DataTable: true,
  Dialog: true,
  Divider: true,
  FAB: true,
  HelperText: true,
  IconButton: true,
  Menu: true,
  Modal: true,
  Portal: true,
  ProgressBar: true,
  RadioButton: true,
  Searchbar: true,
  Snackbar: true,
  Surface: true,
  Switch: true,
  Appbar: true,
  TouchableRipple: true,
  TextInput: true,
  ToggleButton: true,
  Caption: true,
  Headline: true,
  Paragraph: true,
  Subheading: true,
  Title: true,
  Text: true
};
Object.defineProperty(exports, "ActivityIndicator", {
  enumerable: true,
  get: function () {
    return _ActivityIndicator.default;
  }
});
Object.defineProperty(exports, "AnimatedFAB", {
  enumerable: true,
  get: function () {
    return _AnimatedFAB.default;
  }
});
Object.defineProperty(exports, "Appbar", {
  enumerable: true,
  get: function () {
    return _Appbar.default;
  }
});
exports.Avatar = void 0;
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function () {
    return _Badge.default;
  }
});
Object.defineProperty(exports, "Banner", {
  enumerable: true,
  get: function () {
    return _Banner.default;
  }
});
Object.defineProperty(exports, "BottomNavigation", {
  enumerable: true,
  get: function () {
    return _BottomNavigation.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function () {
    return _Button.default;
  }
});
Object.defineProperty(exports, "Caption", {
  enumerable: true,
  get: function () {
    return _Caption.default;
  }
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function () {
    return _Card.default;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function () {
    return _Checkbox.default;
  }
});
Object.defineProperty(exports, "Chip", {
  enumerable: true,
  get: function () {
    return _Chip.default;
  }
});
exports.Colors = void 0;
Object.defineProperty(exports, "DarkTheme", {
  enumerable: true,
  get: function () {
    return _DarkTheme.default;
  }
});
Object.defineProperty(exports, "DataTable", {
  enumerable: true,
  get: function () {
    return _DataTable.default;
  }
});
Object.defineProperty(exports, "DefaultTheme", {
  enumerable: true,
  get: function () {
    return _DefaultTheme.default;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function () {
    return _Dialog.default;
  }
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function () {
    return _Divider.default;
  }
});
exports.Drawer = void 0;
Object.defineProperty(exports, "FAB", {
  enumerable: true,
  get: function () {
    return _FAB.default;
  }
});
Object.defineProperty(exports, "Headline", {
  enumerable: true,
  get: function () {
    return _Headline.default;
  }
});
Object.defineProperty(exports, "HelperText", {
  enumerable: true,
  get: function () {
    return _HelperText.default;
  }
});
Object.defineProperty(exports, "IconButton", {
  enumerable: true,
  get: function () {
    return _IconButton.default;
  }
});
exports.List = void 0;
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function () {
    return _Menu.default;
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function () {
    return _Modal.default;
  }
});
Object.defineProperty(exports, "Paragraph", {
  enumerable: true,
  get: function () {
    return _Paragraph.default;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _Portal.default;
  }
});
Object.defineProperty(exports, "ProgressBar", {
  enumerable: true,
  get: function () {
    return _ProgressBar.default;
  }
});
Object.defineProperty(exports, "Provider", {
  enumerable: true,
  get: function () {
    return _Provider.default;
  }
});
Object.defineProperty(exports, "RadioButton", {
  enumerable: true,
  get: function () {
    return _RadioButton.default;
  }
});
Object.defineProperty(exports, "Searchbar", {
  enumerable: true,
  get: function () {
    return _Searchbar.default;
  }
});
Object.defineProperty(exports, "Snackbar", {
  enumerable: true,
  get: function () {
    return _Snackbar.default;
  }
});
Object.defineProperty(exports, "Subheading", {
  enumerable: true,
  get: function () {
    return _Subheading.default;
  }
});
Object.defineProperty(exports, "Surface", {
  enumerable: true,
  get: function () {
    return _Surface.default;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function () {
    return _Switch.default;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _Text.default;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function () {
    return _TextInput.default;
  }
});
Object.defineProperty(exports, "ThemeProvider", {
  enumerable: true,
  get: function () {
    return _theming.ThemeProvider;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function () {
    return _Title.default;
  }
});
Object.defineProperty(exports, "ToggleButton", {
  enumerable: true,
  get: function () {
    return _ToggleButton.default;
  }
});
Object.defineProperty(exports, "TouchableRipple", {
  enumerable: true,
  get: function () {
    return _TouchableRipple.default;
  }
});
Object.defineProperty(exports, "configureFonts", {
  enumerable: true,
  get: function () {
    return _fonts.default;
  }
});
Object.defineProperty(exports, "overlay", {
  enumerable: true,
  get: function () {
    return _overlay.default;
  }
});
Object.defineProperty(exports, "shadow", {
  enumerable: true,
  get: function () {
    return _shadow.default;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function () {
    return _theming.useTheme;
  }
});
Object.defineProperty(exports, "withTheme", {
  enumerable: true,
  get: function () {
    return _theming.withTheme;
  }
});

var Colors = _interopRequireWildcard(require("./styles/colors"));

exports.Colors = Colors;

var _theming = require("./core/theming");

var _Provider = _interopRequireDefault(require("./core/Provider"));

var _DefaultTheme = _interopRequireDefault(require("./styles/DefaultTheme"));

var _DarkTheme = _interopRequireDefault(require("./styles/DarkTheme"));

var _shadow = _interopRequireDefault(require("./styles/shadow"));

var _overlay = _interopRequireDefault(require("./styles/overlay"));

var _fonts = _interopRequireDefault(require("./styles/fonts"));

var Avatar = _interopRequireWildcard(require("./components/Avatar/Avatar"));

exports.Avatar = Avatar;

var List = _interopRequireWildcard(require("./components/List/List"));

exports.List = List;

var Drawer = _interopRequireWildcard(require("./components/Drawer/Drawer"));

exports.Drawer = Drawer;

var _AnimatedFAB = _interopRequireWildcard(require("./components/FAB/AnimatedFAB"));

Object.keys(_AnimatedFAB).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _AnimatedFAB[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _AnimatedFAB[key];
    }
  });
});

var _Badge = _interopRequireDefault(require("./components/Badge"));

var _ActivityIndicator = _interopRequireDefault(require("./components/ActivityIndicator"));

var _Banner = _interopRequireDefault(require("./components/Banner"));

var _BottomNavigation = _interopRequireDefault(require("./components/BottomNavigation/BottomNavigation"));

var _Button = _interopRequireDefault(require("./components/Button"));

var _Card = _interopRequireDefault(require("./components/Card/Card"));

var _Checkbox = _interopRequireDefault(require("./components/Checkbox"));

var _Chip = _interopRequireDefault(require("./components/Chip"));

var _DataTable = _interopRequireDefault(require("./components/DataTable/DataTable"));

var _Dialog = _interopRequireDefault(require("./components/Dialog/Dialog"));

var _Divider = _interopRequireDefault(require("./components/Divider"));

var _FAB = _interopRequireDefault(require("./components/FAB"));

var _HelperText = _interopRequireDefault(require("./components/HelperText"));

var _IconButton = _interopRequireDefault(require("./components/IconButton"));

var _Menu = _interopRequireDefault(require("./components/Menu/Menu"));

var _Modal = _interopRequireDefault(require("./components/Modal"));

var _Portal = _interopRequireDefault(require("./components/Portal/Portal"));

var _ProgressBar = _interopRequireDefault(require("./components/ProgressBar"));

var _RadioButton = _interopRequireDefault(require("./components/RadioButton"));

var _Searchbar = _interopRequireDefault(require("./components/Searchbar"));

var _Snackbar = _interopRequireDefault(require("./components/Snackbar"));

var _Surface = _interopRequireDefault(require("./components/Surface"));

var _Switch = _interopRequireDefault(require("./components/Switch"));

var _Appbar = _interopRequireDefault(require("./components/Appbar"));

var _TouchableRipple = _interopRequireDefault(require("./components/TouchableRipple/TouchableRipple"));

var _TextInput = _interopRequireDefault(require("./components/TextInput/TextInput"));

var _ToggleButton = _interopRequireDefault(require("./components/ToggleButton"));

var _Caption = _interopRequireDefault(require("./components/Typography/Caption"));

var _Headline = _interopRequireDefault(require("./components/Typography/Headline"));

var _Paragraph = _interopRequireDefault(require("./components/Typography/Paragraph"));

var _Subheading = _interopRequireDefault(require("./components/Typography/Subheading"));

var _Title = _interopRequireDefault(require("./components/Typography/Title"));

var _Text = _interopRequireDefault(require("./components/Typography/Text"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
